File: backend/security-oauth/src/main/java/io/featurehub/web/security/oauth/providers/AzureProvider.java
package io.featurehub.web.security.oauth.providers;

import cd.connect.app.config.ConfigKey;
import cd.connect.app.config.DeclaredConfigResolver;
import io.featurehub.web.security.oauth.AuthClientResult;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;

// register your app here: https://go.microsoft.com/fwlink/?linkid=2083908
// https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow
public class AzureProvider implements OAuth2Provider {
  public static final String PROVIDER_NAME = "oauth2-azure";
  @ConfigKey("oauth2.providers.azure.tenant")
  protected String tenant;
  @ConfigKey("oauth2.providers.azure.secret")
  protected String oauthClientSecret;
  @ConfigKey("oauth2.providers.azure.id")
  protected String oauthClientID;
  @ConfigKey("oauth2.redirectUrl")
  protected String redirectUrl;
  @ConfigKey("oauth2.providers.azure.scopes")
  protected String scopes = "openid email profile";

  private final String actualAuthUrl;
  private final String tokenUrl;

  public AzureProvider() {
    DeclaredConfigResolver.resolve(this);

    actualAuthUrl = String.format("https://login.microsoftonline.com/%s/oauth2/v2.0/authorize?" +
      "response_mode=query&scope=%s&include_granted_scopes=true&response_type=code&client_id=%s&redirect_uri=%s",
      tenant,
      URLEncoder.encode(scopes, StandardCharsets.UTF_8),
      oauthClientID,
      URLEncoder.encode(redirectUrl, StandardCharsets.UTF_8))
      ;

    tokenUrl = String.format("https://login.microsoftonline.com/%s/oauth2/v2.0/token", tenant);
  }

  @Override
  public ProviderUser discoverProviderUser(AuthClientResult authed) {
    Map<String, String> idInfo = Jwt.decodeJwt(authed.getIdToken());

    if (idInfo == null) {
      return null;
    }

    return new ProviderUser.Builder().email(idInfo.get("email")).name(idInfo.get("name")).build();
  }

  @Override
  public String providerName() {
    return PROVIDER_NAME;
  }

  @Override
  public String requestTokenUrl() {
    return tokenUrl;
  }

  @Override
  public String requestAuthorizationUrl() {
    return actualAuthUrl;
  }

  @Override
  public String getClientId() {
    return oauthClientID;
  }

  @Override
  public String getClientSecret() {
    return oauthClientSecret;
  }
}

File: backend/security-oauth/src/main/kotlin/io/featurehub/web/security/oauth/providers/AzureProvider.kt
package io.featurehub.web.security.oauth.providers

import cd.connect.app.config.ConfigKey
import cd.connect.app.config.DeclaredConfigResolver
import io.featurehub.web.security.oauth.AuthClientResult
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

// register your app here: https://go.microsoft.com/fwlink/?linkid=2083908
// https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow
class AzureProvider : OAuth2Provider {
    @ConfigKey("oauth2.providers.azure.tenant")
    protected var tenant: String? = null

    @ConfigKey("oauth2.providers.azure.secret")
    override var clientSecret: String? = null
        protected set

    @ConfigKey("oauth2.providers.azure.id")
    override var clientId: String? = null
        protected set

    @ConfigKey("oauth2.redirectUrl")
    protected var redirectUrl: String? = null

    @ConfigKey("oauth2.providers.azure.scopes")
    protected var scopes = "openid email profile"
    private val actualAuthUrl: String
    private val tokenUrl: String

    override fun discoverProviderUser(authed: AuthClientResult): ProviderUser? {
        val idInfo = Jwt.decodeJwt(authed.idToken) ?: return null
        return ProviderUser.Builder().email(idInfo["email"]).name(idInfo["name"]).build()
    }

    override fun providerName(): String {
        return PROVIDER_NAME
    }

    override fun requestTokenUrl(): String {
        return tokenUrl
    }

    override fun requestAuthorizationUrl(): String {
        return actualAuthUrl
    }

    companion object {
        const val PROVIDER_NAME = "oauth2-azure"
    }

    init {
        DeclaredConfigResolver.resolve(this)
        actualAuthUrl = String.format(
            "https://login.microsoftonline.com/%s/oauth2/v2.0/authorize?" +
                    "response_mode=query&scope=%s&include_granted_scopes=true&response_type=code&client_id=%s&redirect_uri=%s",
            tenant,
            URLEncoder.encode(scopes, StandardCharsets.UTF_8),
            clientId,
            URLEncoder.encode(redirectUrl, StandardCharsets.UTF_8)
        )
        tokenUrl = String.format("https://login.microsoftonline.com/%s/oauth2/v2.0/token", tenant)
    }
}
